---
title: "Turtle annotation stats"
author: "[gwct](https://gwct.github.io/)"
date: "`r format(Sys.time(), '%m/%d/%Y %H:%M:%S %Z')`"
output:
  #html_document:
  rmdformats::robobook:
    highlight: kate
    includes:
      in_header: '../html-chunks/rmd_header.html'
    df_print: paged
    code_folding: hide
    number_sections: true
    toc_depth: 3
---

```{r setup, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

library(ggplot2)
library(dplyr)
library(ggrepel)
library(here)
library(reshape2)
source("../lib/design.r")
```

[< Back to samples](index.html)

Assemblies from each genome were downloaded from NCBI on 07.22.2021

```{r read-data}
annotated_specs = c("Cabingdonii", "Cmydas", "Cpicta", "Cserpentina", "Dcoriacea", "Gagassizi", "Gevgoodei", "Lkempii", "Mmutica", "Mreevesii", "Mterrapin", "Pcastaneus", "Pmegacephalum", "Psinensis", "Tcarolina", "Tscripta")

liftoff_specs = c("Pexpansa", "Esubglobosa", "Mtuberculata", "Dmawii", "Cinsculpta", "Cmccordi", "Rswinhoei", "Amarmorata")
# All species with annotations

final_specs = c(annotated_specs, liftoff_specs)

feature_counts = data.frame()
feature_lens = data.frame()
# Data frames to keep track of feature counts and lengths

for(spec in final_specs){
  #print(spec)
  cur_feature_file = here("data", "feature-counts", paste(spec, "-feature-counts.tab", sep=""))
  cur_feature_counts = read.csv(cur_feature_file, sep="\t", header=T, comment.char="#")
  cur_feature_counts$spec = spec
    
  if(spec %in% annotated_specs){
    cur_feature_counts$type = "Pre-annotated"
  }else{
    cur_feature_counts$type = "Liftoff"
  }
  
  feature_counts = rbind(feature_counts, cur_feature_counts)
  # Read feature counts for the current species
  
  cur_len_file = here("data", "feature-counts", paste(spec, "-feature-counts-lens.tab", sep=""))
  cur_feature_lens = read.csv(cur_len_file, sep="\t", header=T, comment.char="#")
  cur_feature_lens$spec = spec
  
  if(spec %in% annotated_specs){
    cur_feature_lens$type = "Pre-annotated"
  }else{
    cur_feature_lens$type = "Liftoff"
  }
  
  feature_lens = rbind(feature_lens, cur_feature_lens)
  # Read feature lengths for the current species
}

feature_lens$feature[feature_lens$feature=="mRNA"] = "transcript"
# Some files refer to transcripts as "mRNA" so we change the labels here

feature_lens_lt20k = subset(feature_lens, length<=20000)
# A subset of all features less than 20kb long for plotting distributions

```

## Number of genes

```{r gene-fig, out.width="50%", fig.align="center"}
#in_data$Label = factor(in_data$Label, levels=in_data$Label[order(in_data$Coverage, decreasing=T)])
gene_counts = subset(feature_lens, feature=="gene")
gene_counts = gene_counts %>% group_by(spec, type) %>% summarize(count=n(), avg.len=mean(length, na.rm=T))
# Get genes from feature lengths and summarize by species

gene_p = ggplot(gene_counts, aes(x=spec, y=count, color=type)) +
  geom_segment(aes(x=spec, y=0, xend=spec, yend=count), linetype="dotted", color="#666666") +
  geom_point(size=4) +
  scale_y_continuous(limits=c(0,50000), expand=c(0,0)) +
  ylab("# genes") +
  xlab(paste("")) +
  scale_color_manual(values=c("Pre-annotated"=corecol(pal="wilke", numcol=1, offset=1), "Liftoff"=corecol(pal="wilke", numcol=1))) +
  bartheme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, size=10),
        legend.position="bottom",
        legend.margin=margin(t=0, r=0, b=0, l=0, unit="cm"))
print(gene_p)

#cat("Max coverage: ", max(in_data$Coverage), "X\n",
#    "Average coverage: ", signif(mean(in_data$Coverage), 4), "X\n",
#    "Min coverage: ", min(in_data$Coverage), "X\n", sep="")
```

## Distributions of gene length (genes < 20,000bp)

```{r gene-len-dists, out.width="75%", fig.align="center", fig.height=8}
#in_data$Label = factor(in_data$Label, levels=in_data$Label[order(in_data$Coverage, decreasing=T)])
gene_lens = subset(feature_lens_lt20k, feature=="gene")
# Get genes less than 20kb long

gene_dist_p = ggplot(gene_lens, aes(x=length, fill=type)) +
  geom_histogram(bins=50) +
  scale_y_continuous(expand=c(0,0)) + 
  xlab("Gene length") +
  ylab("# genes") +
  facet_wrap(~spec) +
  scale_fill_manual(values=c("Pre-annotated"=corecol(pal="wilke", numcol=1, offset=1), "Liftoff"=corecol(pal="wilke", numcol=1))) +
  bartheme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, size=10),
        legend.position="bottom")
print(gene_dist_p)
```


## Number of transcripts

```{r transcript-fig, out.width="50%", fig.align="center"}
#in_data$Label = factor(in_data$Label, levels=in_data$Label[order(in_data$Coverage, decreasing=T)])
transcript_counts = subset(feature_lens, feature=="transcript")
transcript_counts = transcript_counts %>% group_by(spec, type) %>% summarize(count=n(), avg.len=mean(length, na.rm=T))
# Get CDS and summarize by species

transcript_p = ggplot(transcript_counts, aes(x=spec, y=count, color=type)) +
  geom_segment(aes(x=spec, y=0, xend=spec, yend=count), linetype="dotted", color="#666666") +
  geom_point(size=4) +
  scale_y_continuous(limits=c(0,80000), expand=c(0,0)) +
  ylab("# transcripts") +
  xlab(paste("")) +
  scale_color_manual(values=c("Pre-annotated"=corecol(pal="wilke", numcol=1, offset=1), "Liftoff"=corecol(pal="wilke", numcol=1))) +
  bartheme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, size=10),
        legend.position="bottom")
print(transcript_p)

#cat("Max coverage: ", max(in_data$Coverage), "X\n",
#    "Average coverage: ", signif(mean(in_data$Coverage), 4), "X\n",
#    "Min coverage: ", min(in_data$Coverage), "X\n", sep="")
```

## Distributions of transcript length (genes < 20,000bp)

```{r transcript-len-dists, out.width="75%", fig.align="center", fig.height=8}
transcript_lens = subset(feature_lens_lt20k, feature=="transcript")
# Get transcripts less than 20kb long

transcript_dist_p = ggplot(transcript_lens, aes(x=length, fill=type)) +
  geom_histogram(bins=50) +
  scale_y_continuous(expand=c(0,0)) +
  scale_fill_manual(values=c("Pre-annotated"=corecol(pal="wilke", numcol=1, offset=1), "Liftoff"=corecol(pal="wilke", numcol=1))) +
  xlab("Transcript length") +
  ylab("# transcripts") +
  facet_wrap(~spec) +
  bartheme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, size=10),
        legend.position="bottom")
print(transcript_dist_p)
```

## Genes x transcripts

```{r gene-transcript-fig, out.width="50%", fig.align="center"}

gene_transcript_counts = merge(gene_counts, transcript_counts, by="spec")
# Combine gene and transcript counts into a data frame

gene_transcript_p = ggplot(gene_transcript_counts, aes(x=count.x, y=count.y, color=type.x)) +
  geom_point(size=3, alpha=0.7) +
  geom_smooth(method="glm", se=F, linetype="dashed", color="#333333") +
  geom_text_repel(aes(label=spec)) +
  ylab("# transcripts") +
  xlab("# genes") +
  scale_color_manual(values=c("Pre-annotated"=corecol(pal="wilke", numcol=1, offset=1), "Liftoff"=corecol(pal="wilke", numcol=1))) +
  bartheme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, size=10),
        legend.position="bottom")
print(gene_transcript_p)

```

## Number of coding exons

```{r cds-fig, out.width="50%", fig.align="center"}
#in_data$Label = factor(in_data$Label, levels=in_data$Label[order(in_data$Coverage, decreasing=T)])
cds_counts = subset(feature_lens, feature=="CDS")
cds_counts = cds_counts %>% group_by(spec, type) %>% summarize(count=n(), avg.len=mean(length, na.rm=T))
# Get transcripts and summarize by species

cds_p = ggplot(cds_counts, aes(x=spec, y=count, color=type)) +
  geom_segment(aes(x=spec, y=0, xend=spec, yend=count), linetype="dotted", color="#666666") +
  geom_point(size=4) +
  scale_y_continuous(limits=c(0,1000000), expand=c(0,0)) +
  ylab("# CDS") +
  xlab(paste("")) +
  scale_color_manual(values=c("Pre-annotated"=corecol(pal="wilke", numcol=1, offset=1), "Liftoff"=corecol(pal="wilke", numcol=1))) +
  bartheme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, size=10),
        legend.position="bottom")
print(cds_p)

```

## Distributions of CDS length (CDS < 750bp)

```{r cds-len-dists, out.width="75%", fig.align="center", fig.height=8}
cds_lens = subset(feature_lens_lt20k, feature=="CDS")
cds_lens = subset(cds_lens, length<750)
# Get CDS less than 20kb long

cds_dist_p = ggplot(cds_lens, aes(x=length, fill=type)) +
  geom_histogram(bins=50) +
  scale_y_continuous(expand=c(0,0)) +
  xlab("CDS length") +
  ylab("# CDS") +
  facet_wrap(~spec) +
  scale_fill_manual(values=c("Pre-annotated"=corecol(pal="wilke", numcol=1, offset=1), "Liftoff"=corecol(pal="wilke", numcol=1))) +
  bartheme() +
  theme(axis.text.x=element_text(angle=45, hjust=1, size=10),
        legend.position="bottom")
print(cds_dist_p)
```

## BUSCO results

7480 total BUSCO genes searched

```{r busco-fig, out.width="66%", fig.align="center", fig.height=6}

in_data = read.csv(here("data", "turtles-full.csv"), header=T, comment.char="#")
in_data = subset(in_data, node.type=="tip")
# Read in the turtle data and exclude "alternate" species

busco_data = select(in_data, label, complete.and.single, complete.and.dup, fragmented, missing)
busco_data_long = melt(busco_data, id.vars=c("label"))

busco_p = ggplot(busco_data_long, aes(x=label, y=value, fill=variable, group=label)) +
  geom_bar(stat="identity", position="stack") +
  geom_hline(yintercept=6732, linetype="dashed", col="#333333") +
  scale_y_continuous(expand=c(0,0)) +
  scale_fill_manual(labels=c("Completed+single copy", "Complete+duplicated", "Fragmented", "Missing"), values=corecol(pal="wilke", offset=2)) +
  xlab("") +
  ylab("# of genes") +
  bartheme() +
  theme(legend.position="bottom") +
  guides(fill=guide_legend(nrow=2,byrow=TRUE)) +
  coord_flip()
print(busco_p)

```

### Transcriptome BUSCO


```{r busco-ts-fig, out.width="66%", fig.align="center", fig.height=6}

bt_data = read.csv(here("data", "TurtleAnnotationBuscoScores.csv"), header=T, comment.char="#")
#in_data = subset(in_data, node.type=="tip")
# Read in the turtle data and exclude "alternate" species

busco_data = select(bt_data, Species, CompleteSingleCopy, CompleteDup, Fragmented, Missing)
busco_data_long = melt(busco_data, id.vars=c("Species"))

busco_p = ggplot(busco_data_long, aes(x=Species, y=value, fill=variable, group=Species)) +
  geom_bar(stat="identity", position="stack") +
  geom_hline(yintercept=6732, linetype="dashed", col="#333333") +
  scale_y_continuous(expand=c(0,0)) +
  scale_fill_manual(labels=c("Completed+single copy", "Complete+duplicated", "Fragmented", "Missing"), values=corecol(pal="wilke", offset=2)) +
  xlab("") +
  ylab("# of genes") +
  bartheme() +
  theme(legend.position="bottom") +
  guides(fill=guide_legend(nrow=2,byrow=TRUE)) +
  coord_flip()
print(busco_p)

```

## Liftoff annotations

[Liftoff](https://github.com/agshumate/Liftoff) is a program that maps annotated features from one genome to another.

### Genes

```{r liftoff-genes, out.width="50%", fig.align="center", fig.height=6}

liftoff = data.frame("spec"=c(), "genes"=c(), "transcripts"=c(), "cds"=c(), "ref"=c(), "ref.genes"=c(), "ref.transcripts"=c())
for(spec in liftoff_specs){
  #print(spec)
  
  row = subset(in_data, label==spec)
  ref = row$liftoff.ref
  ref_genes = in_data[in_data$label==ref,]$num.genes
  ref_transcripts = in_data[in_data$label==ref,]$num.transcripts

  
  liftoff = rbind(liftoff, data.frame("spec"=spec, "genes"=row$num.genes, "transcripts"=row$num.transcripts, "ref"=ref, "ref.genes"=ref_genes, "ref.transcripts"=ref_transcripts))
}

liftoff$perc.ref.genes = liftoff$genes / liftoff$ref.genes
liftoff$genes.label = paste(liftoff$genes, " / ", liftoff$ref.genes, " (", liftoff$ref, ")", sep="")

liftoff_genes_p = ggplot(liftoff, aes(x=spec, y=perc.ref.genes)) +
  geom_bar(stat="identity", position="stack", fill=corecol(numcol=1, pal="wilke", offset=2)) +
  geom_hline(yintercept=0.9, linetype="dashed", col="#333333") +
  geom_text(aes(label=genes.label), size=4, position=position_stack(vjust=0.5), color="#f2f2f2") +
  scale_y_continuous(expand=c(0,0)) +
  #scale_fill_manual(labels=c("Completed+single copy", "Complete+duplicated", "Fragmented", "Missing"), values=corecol(pal="wilke", offset=2)) +
  xlab("") +
  ylab("Proportion reference genes recovered") +
  bartheme() +
  theme(legend.position="bottom") +
  guides(fill=guide_legend(nrow=2,byrow=TRUE)) +
  coord_flip()
print(liftoff_genes_p)

```

### Transcripts

```{r liftoff-transcripts, out.width="50%", fig.align="center", fig.height=6}

liftoff$perc.ref.transcripts = liftoff$transcripts / liftoff$ref.transcripts
liftoff$transcripts.label = paste(liftoff$transcripts, " / ", liftoff$ref.transcripts, " (", liftoff$ref, ")", sep="")

liftoff_transcripts_p = ggplot(liftoff, aes(x=spec, y=perc.ref.transcripts)) +
  geom_bar(stat="identity", position="stack", fill=corecol(numcol=1, pal="wilke", offset=6)) +
  geom_hline(yintercept=0.9, linetype="dashed", col="#333333") +
  geom_text(aes(label=genes.label), size=4, position=position_stack(vjust=0.5), color="#f2f2f2") +
  scale_y_continuous(expand=c(0,0)) +
  xlab("") +
  ylab("Proportion reference transcripts recovered") +
  bartheme() +
  theme(legend.position="bottom") +
  guides(fill=guide_legend(nrow=2,byrow=TRUE)) +
  coord_flip()
print(liftoff_transcripts_p)

```

## Toga annotations

[Toga](https://github.com/hillerlab/TOGA) is a program that maps annotated features from one genome to another using a whole genome alignment.

Out of 26,849 genes annotated in Dcoriacea, 18,331 were used as input to Toga.

```{r toga-genes, out.width="50%", fig.align="center", fig.height=6}

toga_file = here("data", "dcoriacea-toga.tsv")
toga_data = read.csv(toga_file, header=T, sep="\t")
# Read the toga data

toga_counts = NULL
# An empty var to store the counts for each species

toga_classes = c("I", "PI", "M", "PM", "L", "UL", "PG", "abs")
toga_class_names = c("Intact", "Partially intact", "Missing", "Partially missing", "Lost", "Uncertain loss","Paralogous", "absent")
cols = corecol(numcol=length(toga_classes))
names(cols)=toga_classes
# Abbreviations and names for the toga classes, as well as colors for the plot

for(i in 1:length(names(toga_data))){
  if(i == 1){
    names(toga_data)[i] = "gene.id"
    # For the first column, add the name
  } else {
    spec = strsplit(names(toga_data)[i], ".", fixed=T)[[1]][1]
    names(toga_data)[i] = spec
    # Extract the species name from the current header and change the header
    
    cur_counts = table(toga_data[spec])[toga_classes]
    cur_counts = as.data.frame(cur_counts)
    cur_counts$spec = spec
    # Count each class for the current species
    
    if(is.null(toga_counts)){
      toga_counts = cur_counts
    } else {
      toga_counts = rbind(toga_counts, cur_counts)
    }
    # Add the counts to the count df
  }
}

toga_counts$Var1 = factor(toga_counts$Var1, levels=toga_classes)

toga_p = ggplot(toga_counts, aes(x=spec, y=Freq, fill=Var1, group=spec)) +
  geom_bar(stat="identity", position="stack") +
  geom_hline(yintercept=0.9*nrow(toga_data), linetype="dashed", col="#d3d3d3") +
  scale_y_continuous(expand=c(0,0)) +
  scale_fill_manual(labels=toga_class_names, values=cols) +
  xlab("") +
  ylab("# of genes") +
  bartheme() +
  theme(legend.position="bottom") +
  guides(fill=guide_legend(nrow=2,byrow=TRUE)) +
  coord_flip()
print(toga_p)


```



















[< Back to samples](index.html)