---
title: "Turtle trees"
author: "[gwct](https://gwct.github.io/)"
date: "`r format(Sys.time(), '%m/%d/%Y %H:%M:%S %Z')`"
navbar:
  title: Murinae seq
  right:
    - text: "Home"
      href: https://goodest-goodlab.github.io/murinae-seq/
    - text: "210 exomes"
      href: https://goodest-goodlab.github.io/murinae-seq/summary_210.html
output:
  #html_document:
  rmdformats::robobook:
    highlight: kate
    includes:
      in_header: '../html-chunks/rmd_header.html'
    df_print: paged
    code_folding: hide
    number_sections: true
    toc_depth: 3
---

```{r setup, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

library(ggplot2)
library(cowplot)
library(ggbeeswarm)
library(dplyr)
#library(kableExtra)
library(tidyr)
library(ggtree)
library(phytools)
library(phangorn)
#library(reshape2)
#library(ggExtra)
library(ggrepel)
#library(vroom)
#library(ggdist)
library(here)
#source("C:/bin/core/r/design.r")
#source("C:/bin/core/r/get_tree_info.r")
source("C:/Users/grt814/bin/core/r/design.r")
source("C:/Users/grt814/bin/core/r/get_tree_info.r")

#htmltools::includeHTML("../html-chunks/rmd_nav.html")
```

[< Back to summary](index.html)

# Species tree topology

```{r read, out.width="100%", fig.align = "center", warning=FALSE}
tree_type = "topo"
save_tree_fig = F

if(tree_type == "astral"){
  cat("188 species.\n11,775 coding loci.\nGene trees inferred with IQtree.\nSpecies tree inferred with ASTRAL (no branch lengths shown).\n")
}else if(tree_type == "concat"){
  cat("188 species.\n11,775 coding loci.\nGene trees inferred with IQtree.\nSpecies tree inferred by concatenation of all loci with IQtree.\n")
}else if(tree_type == "topo"){
  cat("Topology extracted from Thomson et al. 2021: https://www.pnas.org/content/118/7/e2012215118")
}
# Info for different tree input types

# tree_file = "../../data/trees/full_coding_iqtree_astral.cf.rooted.tree"
# astral_tree = read.tree(tree_file)
# 
# tree_to_df_list = treeToDF(astral_tree)
# tree_info_astral = tree_to_df_list[["info"]]
# #tree_info_astral = treeToDF(astral_tree)
# tree_info_astral = tree_info_astral %>% separate(label, c("astral", "gcf", "scf"), sep="/", remove=F)
# tree_info_astral$astral[tree_info_astral$node.type=="tip"] = NA
# tree_info_astral$astral = as.numeric(tree_info_astral$astral)
# tree_info_astral$gcf = as.numeric(tree_info_astral$gcf)
# tree_info_astral$scf = as.numeric(tree_info_astral$scf)
# Read astral tree data

# concat_file = "../../data/trees/full_coding_iqtree_concat.cf.rooted.tree"
# concat_tree = read.tree(concat_file)
# 
# tree_to_df_list = treeToDF(concat_tree)
# tree_info_concat = tree_to_df_list[["info"]]
# #tree_info_concat = treeToDF(concat_tree)
# tree_info_concat = tree_info_concat %>% separate(label, c("bootstrap", "gcf", "scf"), sep="/", remove=F)
# tree_info_concat$bootstrap[tree_info_concat$node.type=="tip"] = NA
# tree_info_concat$bootstrap = as.numeric(tree_info_concat$bootstrap)
# tree_info_concat$gcf = as.numeric(tree_info_concat$gcf)
# tree_info_concat$scf = as.numeric(tree_info_concat$scf)
# Read concat tree data

topo_file = here("data", "tree", "turtle-samples-topo-rooted.tre")
topo_tree = read.tree(topo_file)

tree_to_df_list = treeToDF(topo_tree)
tree_info_topo = tree_to_df_list[["info"]]
# Manual topo tree -- Need to fix this for no branch lengths!!

topo_file_used = here("data", "tree", "turtle-samples-topo-rooted-included.tre")
topo_tree_used = read.tree(topo_file_used)

tree_to_df_list = treeToDF(topo_tree_used)
tree_info_topo_used = tree_to_df_list[["info"]]

#rf = RF.dist(concat_tree, astral_tree)
#nrf = RF.dist(concat_tree, astral_tree, normalize=T)

#treecomp = comparePhylo(concat_tree, astral_tree, plot=T)
# Stores node ids of common clades between trees!

#write.csv(tree_info, "../../data/trees/full-coding-concat-cf-rooted.csv", row.names=F)

if(tree_type == "astral"){
  tree_info = tree_info_astral
  turtle_tree = astral_tree
  xmax = 31
  iq_tree_labels = "../../data/trees/full_coding_iqtree_astral.cf.branch.rooted"
  cf_stat_file = "../../data/trees/full_coding_iqtree_astral.cf.stat"
  cf_rep_dir = "../../data/trees/astral-cf-reps/"
  delta_outfile = "../../data/trees/astral-delta.tab"
}else if(tree_type == "concat"){
  tree_info = tree_info_concat
  turtle_tree = concat_tree
  xmax = 0.125
  iq_tree_labels = "../../data/trees/full_coding_iqtree_concat.cf.branch.rooted"
  cf_stat_file = "../../data/trees/full_coding_iqtree_concat.cf.stat"
  cf_rep_dir = "../../data/trees/concat-cf-reps/"
  delta_outfile = "../../data/trees/concat-delta.tab"
}else if(tree_type == "topo"){
  tree_info = tree_info_topo
  turtle_tree = topo_tree
  xmax = 12
}
# Set some values based on tree type

#cf_stats = read.table(cf_stat_file, header=T)
```

```{r read-sample-data}
in_data_file = here("data", "turtles.csv")
in_data = read.csv(in_data_file, header=T, comment.char="#")

full_data = merge(in_data, tree_info, by="label", all=T)
#in_data = subset(in_data, Alternate=="N")

full_data = full_data %>% arrange(node)

include_data = subset(in_data, !label %in% c("Tcarolina", "Gagassizi", "Camboinensis"))
include_data = merge(include_data, tree_info_topo_used, by="label", all=T)
include_data = include_data %>% arrange(node)


# in_data$num.genes = NA
# in_data$avg.gene.len = NA
# in_data$num.genes.lt20k = NA
# in_data$avg.gene.len.lt20k = NA
# in_data$num.transcripts = NA
# in_data$avg.transcript.len = NA
# in_data$num.transcripts.lt20k = NA
# in_data$avg.transcript.len.lt20k = NA

# Read in the input data
```

## All available species

```{r topology-fig, out.width="75%", fig.align = "center", warning=FALSE, fig.height=10}
h = corecol(numcol=1, pal="wilke", offset=3)
l = corecol(numcol=1, offset=3)
# Colors

famcol = corecol(pal="trek", numcol=1, offset=2)
famoffset = 2.1

gcf_tree = ggtree(turtle_tree, size=0.8, ladderize=F) +#, aes(color=full_data$Family)) +
  xlim(0, xmax) +
  geom_tiplab(aes(color=full_data$Include), fontface='italic', size=4) +
  scale_color_manual(name='Include?', values=c("N"=corecol(pal="wilke", numcol=1,offset=1), "Y"=corecol(numcol=1, offset=1))) +
  #geom_text(aes(label=node), hjust=-.3, vjust=-.3, color="#ff6db6") +
  
  geom_cladelabel(node=25, label="Podocnemididae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=24, label="Pelomedusidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=48, label="Chelidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=21, label="Carettochelyidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=46, label="Trionychidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=18, label="Dermochelyidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=44, label="Cheloniidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=15, label="Chelydridae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=14, label="Dermatemydidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=13, label="Platysternidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=37, label="Emydidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=34, label="Testudinidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=31, label="Geomydidae", align=T, color=famcol, offset=famoffset) +
  
  theme(legend.position=c(0.05,0.2))
print(gcf_tree)

if(save_tree_fig){
  gcf_tree = gcf_tree + geom_text(aes(x=branch, label=ifelse(tree_info$node.type=="internal",as.character(node), ''), label.size=NA, fill="transparent"), size=2, vjust=-0.2)
  tree_outfile = paste("../../data/trees/", tree_type, "-gcf-tree.pdf", sep="")
  ggsave(tree_outfile, gcf_tree, width=8, height=16, unit="in")
}
# Tree
```

## Selected species

```{r selected-fig, out.width="75%", fig.align = "center", warning=FALSE, fig.height=10}
h = corecol(numcol=1, pal="wilke", offset=3)
l = corecol(numcol=1, offset=3)
# Colors

famcol = corecol(pal="trek", numcol=1, offset=2)
famoffset = 2.1

gcf_tree = ggtree(topo_tree_used, size=0.8, ladderize=F) +#, aes(color=full_data$Family)) +
  xlim(0, xmax) +
  geom_tiplab(aes(color=include_data$Annotation.file), fontface='italic', size=4) +
  scale_color_manual(name='Annotated?', values=c("N"=corecol(pal="wilke", numcol=1), "Y"=corecol(pal="wilke", numcol=1, offset=1))) +
  geom_text(aes(label=node), hjust=-.3, vjust=-.3, color="#ff6db6") +
  
  geom_cladelabel(node=22, label="Podocnemididae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=21, label="Pelomedusidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=42, label="Chelidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=18, label="Carettochelyidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=40, label="Trionychidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=15, label="Dermochelyidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=38, label="Cheloniidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=12, label="Chelydridae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=11, label="Dermatemydidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=10, label="Platysternidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=32, label="Emydidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=30, label="Testudinidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=28, label="Geomydidae", align=T, color=famcol, offset=famoffset) +
  
  theme(legend.position=c(0.05,0.2))
print(gcf_tree)

if(save_tree_fig){
  gcf_tree = gcf_tree + geom_text(aes(x=branch, label=ifelse(tree_info$node.type=="internal",as.character(node), ''), label.size=NA, fill="transparent"), size=2, vjust=-0.2)
  tree_outfile = paste("../../data/trees/", tree_type, "-gcf-tree.pdf", sep="")
  ggsave(tree_outfile, gcf_tree, width=8, height=16, unit="in")
}
# Tree
```


```{r read-annotation-data, out.width="75%", fig.align = "center", warning=FALSE, eval=T}

annotated_specs = c("Cabingdonii", "Cmydas", "Cpicta", "Cserpentina", "Dcoriacea", "Gagassizi", "Gevgoodei", "Lkempii", "Mmutica", "Mreevesii", "Mterrapin", "Pcastaneus", "Pmegacephalum", "Psinensis", "Tcarolina", "Tscripta")

liftoff_specs = c("Pexpansa", "Esubglobosa", "Mtuberculata", "Dmawii", "Cinsculpta", "Cmccordi", "Rswinhoei", "Amarmorata")
# All species with annotations

final_specs = c(annotated_specs, liftoff_specs)

feature_counts = data.frame()
feature_lens = data.frame()

for(spec in final_specs){
  cur_feature_file = here("data", "feature-counts", paste(spec, "-feature-counts.tab", sep=""))
  cur_feature_counts = read.csv(cur_feature_file, sep="\t", header=T, comment.char="#")
  cur_feature_counts$spec = spec
  feature_counts = rbind(feature_counts, cur_feature_counts)
  # Read feature counts for the current species
  
  cur_len_file = here("data", "feature-counts", paste(spec, "-feature-counts-lens.tab", sep=""))
  cur_feature_lens = read.csv(cur_len_file, sep="\t", header=T, comment.char="#")
  cur_feature_lens$spec = spec
  feature_lens = rbind(feature_lens, cur_feature_lens)
  # Read feature lengths for the current species
}
# Read in the annotation data

```

```{r add-feature-counts, out.width="75%", fig.align = "center", warning=FALSE, eval=T}

for(cur_spec in final_specs){
  #print(cur_spec)

  cur_genes = subset(feature_lens, spec==cur_spec & feature=="gene")
  cur_transcripts = subset(feature_lens, spec==cur_spec)
  cur_transcripts = subset(cur_transcripts, feature=="transcript" | feature=="mRNA")
  cur_transcripts$feature = "transcript"
  # Subset the genes and transcripts, combining "transcript" and "mRNA" features as transcripts
  
  in_data$num.genes[in_data$label==cur_spec] = nrow(cur_genes)
  in_data$avg.gene.len[in_data$label==cur_spec] = mean(cur_genes$length, na.rm=T)
  # Genes
  
  cur_genes_lt20k = subset(cur_genes, length <= 20000)
  in_data$num.genes.lt20k[in_data$label==cur_spec] = nrow(cur_genes_lt20k)
  in_data$avg.gene.len.lt20k[in_data$label==cur_spec] = mean(cur_genes_lt20k$length, na.rm=T)
  # Genes less than 20kb
  
  in_data$num.transcripts[in_data$label==cur_spec] = nrow(cur_transcripts)
  in_data$avg.transcript.len[in_data$label==cur_spec] = mean(cur_transcripts$length, na.rm=T)
  # Transcripts

  cur_transcripts_lt20k = subset(cur_transcripts, length <= 20000)
  in_data$num.transcripts.lt20k[in_data$label==cur_spec] = nrow(cur_transcripts_lt20k)
  in_data$avg.transcript.len.lt20k[in_data$label==cur_spec] = mean(cur_transcripts_lt20k$length, na.rm=T)
  # Transcripts less than 20kb
}
# For each annotated species, add the feature counts and average lengths to the main data frame

```

```{r merge-tree-data, out.width="75%", fig.align = "center", warning=FALSE}

full_data = merge(in_data, tree_info, by="label", all=T)
full_data = full_data %>% arrange(node)
full_data$label[full_data$label==""]=NA
write.csv(full_data, here("data", "turtles-full.csv"), row.names=F)
# Merge the main data frame and the tree data frame and write to file

```

## Selected species with branch lengths

```{r get-bl, out.width="75%", fig.align = "center", warning=FALSE, fig.height=10}

con_tree_file = here("data", "tree", "thomson-trees", "turtle.con.tre")
# Tree from Thomson et al. with branch lengths

con_tree = read.nexus(con_tree_file)
con_tree_rooted = root(con_tree, c("Alligator_mississippiensis", "Gallus_gallus"))
tree_to_df_list = treeToDF(con_tree_rooted)
con_info = tree_to_df_list[["info"]]
# Read, root, and parse the tree

tips_to_keep = data.frame("current.tip"=c("Cmccordi", "Mmutica", "Mreevesii", "Cabingdonii", "Gevgoodei", "Tscripta", "Mterrapin", "Cpicta", "Amarmorata", "Pmegacephalum", "Dmawii", "Cserpentina", "Cmydas", "Lkempii", "Dcoriacea", "Psinensis", "Rswinhoei", "Cinsculpta", "Esubglobosa", "Mtuberculata", "Pcastaneus", "Pexpansa"), 
                          "thomson.tip"=c("Cuora_mccordi__HBS41892", "Mauremys_mutica__MVZ230476", "Mauremys_reevesii_HBS113152", "Chelonoidis_abingdoni_HBS117543", "Gopherus_agassizii_GOAG01", "Trachemys_scripta_elegans_HBS23001", "Malaclemys_terrapin_MVZ137745", "Chrysemys_picta_HBS27134", "Emys_marmorata_HBS39734", "Platysternon_megacephalum_HBS16255", "Dermatemys_mawii_HBS116865", "Chelydra_serpentina_HBS23551", "Chelonia_mydas_HBS109887", "Lepidochelys_olivacea_HBS125086", "Dermochelys_coriacea__MVZ149847", "Pelodiscus_sinensis_AMNH106666", "Rafetus_swinhoei_sp_AMNH200163", "Carretochelys_insculpta_HBS119093", "Emydura_subglobosa_244", "Mesoclemmys_gibbus_BPN1497", "Pelusios_castaneus_CAS219222", "Podocnemis_expansa_HBS117929"))
# Get the tips in the Thomson tree that correspond to the tips in our tree

new_tree = keep.tip(con_tree_rooted, tips_to_keep$thomson.tip)
# Prune the Thomson tree to our tips

for(i in 1:length(new_tree$tip.label)){
  #print(new_tree[["tip.label"]][i])
  cur_tip = new_tree[["tip.label"]][i]
  new_tree[["tip.label"]][i] = tips_to_keep$current.tip[tips_to_keep$thomson.tip == cur_tip]
  #print(new_tree[["tip.label"]][i])
}
# For each tip in the new pruned tree, re-name the tip to match our genome abbr.

tree_to_df_list = treeToDF(new_tree)
new_info_full = tree_to_df_list[["info"]]
new_info_full = merge(in_data, new_info_full, by="label", all.y=T)
new_info_full = new_info_full %>% arrange(node)
new_info_full$label[new_info_full$label==""]=NA

include_tips = subset(include_data, node.type=="tip")
# Read and parse the new pruned tree

new_info = merge(new_info_full, select(include_tips, label, Annotation.file), by="label", all.x=T)
# Merge the tip info from the new tree and the original tree

new_info = new_info[order(new_info$node), ]
# Re-sort the data frame by R node order after the merge so the trees still work

write.csv(new_info, here("data", "turtles-include.csv"), row.names=F)
# Merge the main data frame and the tree data frame and write to file

famoffset = 0.05

new_tree_p = ggtree(new_tree, size=0.8, ladderize=F) +#, aes(color=full_data$Family)) +
  xlim(0, 0.3) +
  geom_tiplab(aes(color=new_info$Annotation.file), fontface='italic', size=4) +
  scale_color_manual(name='Annotated?', values=c("N"=corecol(pal="wilke", numcol=1), "Y"=corecol(pal="wilke", numcol=1, offset=1))) +
  geom_text(aes(label=node), hjust=-.3, vjust=-.3, color="#ff6db6") +
  
  geom_cladelabel(node=18, label="Podocnemididae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=17, label="Pelomedusidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=42, label="Chelidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=2, label="Carettochelyidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=26, label="Trionychidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=7, label="Dermochelyidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=40, label="Cheloniidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=6, label="Chelydridae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=8, label="Dermatemydidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=19, label="Platysternidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=34, label="Emydidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=30, label="Testudinidae", align=T, color=famcol, offset=famoffset) +
  geom_cladelabel(node=31, label="Geomydidae", align=T, color=famcol, offset=famoffset) +

  theme(legend.position=c(0.05,0.1))

  rotate <- ggtree::rotate
  new_tree_p = new_tree_p %>% rotate(43) %>% rotate(42) %>% rotate(24) %>% rotate(27) %>% rotate(40) %>% rotate(38) %>% rotate(25) %>% rotate(35) %>% rotate(29) %>% rotate(30) %>% rotate(31)
  # Rotate nodes in the new tree to match the ordering of the original trees
  
print(new_tree_p)

new_tree$node.label = NULL
# No branch labels for cactus

write.tree(new_tree, file(here("data", "tree", "pruned-thomson.tre")))
# Write the new tree out

```


[< Back to summary](index.html)/